### DISCUSSION TITLE
Project Roadmap

### DISCUSSION THREAD

**Discussion by dlrudie** [dlrudie/Snip#469] [General]
**READ FIRST**
This is _**not**_ a discussion page for feature requests or bugs. Continue to submit tickets for feature requests, bugs, etc. This discussion is for developers and contributors to help shape Snip.

Project Contributors,

Thank you for all your help with translations, bug reports, and replying to issues. I appreciate the assistance and help you provide.

I am rewriting/reworking Snip. I'd like some feedback from developers on what you'd like to see to make development/growth/maintenance of the project.

I get a lot of requests for a macOS and Linux version of Snip, as well as requests for supporting a lot of other players. I don't have the time to make all of that a reality. However, going forward I can make it easier for other people to implement those.

To start with I'm going to split Snip apart. The first part of it will be libSnip which will encompass core functionality and be cross-platform. libSnip will also implement a plugin interface. Spotify, iTunes, etc. will be plugins. The main GUI for Snip will also be separate now as well. This will allow a native implementation on other operating systems. For example, in Linux you would reference libSnip and you could make an interface out of Qt, GTK, or whatever else is out there. macOS could have a native interface as well. This will also open up the possibility to implement libSnip in another application as well.

As for the plugins you would just reference libSnip and then override functions to implement the player of choice. This should allow for developers to not have to download or mess with all of the code for Snip. All they need is libSnip.dll loaded into their project and they can write the code for the player of choice. This will also untie the player support from libSnip directly as well. When registering a plugin it should be registered as to which operating systems it's compatible with as well. iTunes support obviously won't work on Linux. I also am unsure if I could even implement iTunes support in macOS through C#.

Aside from that here's a very rough/crude outline of features I'd like to implement going forward (these are not in any particular order whatsoever). I'll try to mention the feature and then a description of why.
* No matter which plugins are loaded on first-time-setup launch of Snip it will ask you to select which plugin/player you'd like to use. This will be useful for people that want to use Snip for iTunes and not have to deal with Spotify asking to log in.
* Settings will be saved in a JSON file instead of the registry. This will be self-contained now to help with cross-platform support.
* Snip GUI for output formatting will have a much easier to use. It will also have a preview output section as well. More on this further down in this list.
* Snip will use an in-app browser window to authorize spotify instead of opening in the default browser. This will allow automatic control of the window and once logged in I can close the window instead of requiring the user to manually close their browser window each time.
* Get rid of the local webserver. I've registered a domain I plan to use for Snip in the future as well as handling logins. This should provide a more reliable endpoint.
* No longer require Snip to be run as admin (this is never a good idea, honestly).
* I'd like to modernize the GUI. Currently Snip does not handle high DPI displays well.
* Retain all the same features as before and add more. Set length of time to keep metadata before it's redownloaded. This should help prevent stale data.
* Output not just to Snip.txt but also to Snip.json. I don't know why but also why not?
* Display current album artwork of playing song within interface, as well as song information.
* Add lots of accessibility support. Currently Snip supports multiple languages for the interface but we can go further. I'd like to implement other accessibility features such as text scaling for the interface for readability, for output formatting add an option to use drag&drop bubbles instead of text (will support touch interface as well), voiceover support, and whatever else I can research that will help everyone enjoy Snip.
* Better error catching/reporting/logging.
* The Spotify API provides TOO MUCH data and I love data. Maybe implement some of this stuff. Track play count? Top genres? Current playing time and track length? Top skipped tracks? Who knows?

Again, this is not a page for feature requests or bug reports, but you may discuss features/bugs.

**Comment by murko69** [comment]
I'm a total statistics nerd and it kinda tickle my senses when you speak of the 'massive API' and the fact that you want to dive in :)
Are you thinking just for the interface or as a possible output as well (talking about your examples: "Track play count? Top genres? Current playing time and track length? Top skipped tracks? Who knows?")? 
Maybe as a separate statistic output?

'Most interesting year' sounds like nice nerdy one I personally would enjoy :P 

I'm guessing more possible outputs than today 'track, artist, album, trackID' would also be made possible directly in GUI? More album info for example, like the ones already in Snip_Metadata.json (release date, year, number of tracks for example).
Is country (of artist/band) presented in the API and possible to retrieve?

I'm just spitballing here - not requesting :)

**Reply by dlrudie** [reply]
Both interface and output. It wouldn't take much to dump it out to a text/json file that can be parsed however anyone wants to use it.

It's been awhile since I fully looked at one of the metadata files but anything within that file can be saved separately.

I haven't fully went through all of the Spotify API yet as I've mostly been working on Snip itself for now. Once I get to the Spotify plugin I planned to dig more into it. I figure the data is there so why not take advantage of what's there.

https://developer.spotify.com/documentation/web-api/reference/

To be honest I have no idea how people would use the additional information. I'd like to include some of the more common/wanted statistics into the interface that people can see. Anything extra can be dumped to a text/json file for people to parse.

Snip originally started out as something I wrote for myself to display the music I was listening to while streaming on Twitch. With Twitch cracking down on DMCA violations, and providing their own music player, I feel like Snip's purpose has become less needed. I'd like to add additional features/support to make it a little bit more relevant. It'd be nice if the Spotify API would include the key/value for royalty-free/public domain music Snip might be able to automatically skip the song if not or notify the user.

**Comment by dlrudie** [comment]
I've got an initial base library in progress and, for the sake of testing on multiple platforms, have started a simple CLI version of Snip. I've only started with Spotify and it's not fully working yet.

Development has shifted away from using Visual Studio itself and everything is being developed with Visual Studio Code instead. I may still run everything through Visual Studio for debugging/optimizations but that will be on my part and no one else should need Visual Studio to build it.

Libraries used are official Microsoft .NET and not using mono.
